File: c/day17.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#define MAX_OUTPUT_LENGTH 4096
#define INSTRUCTION_SIZE 2

typedef struct
{
    int64_t reg_a;
    int64_t reg_b;
    int64_t reg_c;
    int64_t program[16];
    size_t program_length;
} VM_State;

static inline int64_t find_combo(const VM_State *state, int64_t operand)
{
    switch (operand)
    {
    case 0:
    case 1:
    case 2:
    case 3:
        return operand;
    case 4:
        return state->reg_a;
    case 5:
        return state->reg_b;
    case 6:
        return state->reg_c;
    default:
        fprintf(stderr, "Error! Invalid operand: %lld\n", operand);
        return -1;
    }
}

static inline int64_t op0(VM_State *state, int64_t operand)
{
    int64_t to_power = find_combo(state, operand);
    return state->reg_a >> to_power;
}

static inline int64_t op1(VM_State *state, int64_t operand)
{
    return state->reg_b ^ operand;
}

static inline int64_t op2(VM_State *state, int64_t operand)
{
    return find_combo(state, operand) & 0x7;
}

static inline int op3(VM_State *state, int64_t operand, size_t *ip)
{
    if (state->reg_a == 0)
    {
        return 0;
    }
    *ip = operand;
    return 1;
}

static inline int64_t op4(VM_State *state, int64_t operand)
{
    return state->reg_b ^ state->reg_c;
}

static inline int64_t op5(VM_State *state, int64_t operand)
{
    return find_combo(state, operand) & 0x7;
}

static inline int64_t op6(VM_State *state, int64_t operand)
{
    int64_t to_power = find_combo(state, operand);
    return state->reg_a >> to_power;
}

static inline int64_t op7(VM_State *state, int64_t operand)
{
    int64_t to_power = find_combo(state, operand);
    return state->reg_a >> to_power;
}

char *run_process(VM_State *state)
{
    if (!state)
        return NULL;

    char *output = malloc(MAX_OUTPUT_LENGTH);
    if (!output)
        return NULL;

    char *current_pos = output;
    size_t remaining_space = MAX_OUTPUT_LENGTH - 1;
    size_t ip = 0;

    while (ip < state->program_length)
    {
        int64_t opcode = state->program[ip];
        int64_t operand = state->program[ip + 1];
        int jump_occurred = 0;

        switch (opcode)
        {
        case 0:
            state->reg_a = op0(state, operand);
            break;
        case 1:
            state->reg_b = op1(state, operand);
            break;
        case 2:
            state->reg_b = op2(state, operand);
            break;
        case 3:
            jump_occurred = op3(state, operand, &ip);
            if (!jump_occurred)
            {
                ip += INSTRUCTION_SIZE;
            }
            continue;
        case 4:
            state->reg_b = op4(state, operand);
            break;
        case 5:
        {
            int64_t result = op5(state, operand);
            int written = snprintf(current_pos, remaining_space, "%lld,", result);
            if (written > 0 && written < remaining_space)
            {
                current_pos += written;
                remaining_space -= written;
            }
            break;
        }
        case 6:
            state->reg_b = op6(state, operand);
            break;
        case 7:
            state->reg_c = op7(state, operand);
            break;
        }

        if (!jump_occurred)
        {
            ip += INSTRUCTION_SIZE;
        }
    }

    // Remove trailing comma if exists
    if (current_pos > output && *(current_pos - 1) == ',')
    {
        *(current_pos - 1) = '\0';
    }

    return output;
}

int main()
{
    VM_State state = {
        .reg_b = 0,
        .reg_c = 0,
        .program = {2, 4, 1, 1, 7, 5, 1, 5, 4, 1, 5, 5, 0, 3, 3, 0},
        .program_length = 16};

    const char *target_output = "2,4,1,1,7,5,1,5,4,1,5,5,0,3,3,0";
    const int64_t target_value = 999999117440LL; // You can make this much larger now

    printf("Starting search up to %lld\n", target_value);

    for (int64_t i = 0; i <= target_value; i++)
    {
        // Print progress every 5000 iterations
        if (i % 500000 == 0)
        {
            printf("Progress: %lld / %lld\n", i, target_value);
        }

        // Set up state for this iteration
        state.reg_a = i;
        state.reg_b = 0;
        state.reg_c = 0;

        // Run the program
        char *output = run_process(&state);
        if (!output)
        {
            fprintf(stderr, "Error: Failed to get output\n");
            continue;
        }

        // Check if we found a match
        if (strcmp(output, target_output) == 0)
        {
            printf("\nFound matching value!\n");
            printf("Register A value: %lld\n", i);
            printf("Output: %s\n", output);
            free(output);
            break;
        }

        free(output);
    }

    return 0;
}

File: python/day17.py

import re


def read_file() -> list:
    with open('../data/day17_small.txt', 'r') as file:
        file_content = file.read()

        in_map = []
        lines = file_content.strip().split('\n')

        # Process three lines at a time
        for i in range(0, len(lines)):

            # Extract numbers using regex
            reg_a = re.findall(
                r'Register A: (\d+)', lines[i])

            reg_b = re.findall(
                r'Register B: (\d+)', lines[i])

            reg_c = re.findall(
                r'Register C: (\d+)', lines[i])

            prog = re.findall(
                # r'Program: ', lines[i])
                r'Program: (\d+(,\d+)*)', lines[i])

            if len(reg_a):
                in_map.append(int(reg_a[0]))
            elif len(reg_b):
                in_map.append(int(reg_b[0]))
            elif len(reg_c):
                in_map.append(int(reg_c[0]))
            elif len(prog):
                in_map.append([int(i) for i in list(prog[0][0])[::2]])

    return in_map


def find_combo(reg_a, reg_b, reg_c, operand: int):

    if operand in [0, 1, 2, 3]:
        return operand
    elif operand == 4:
        return reg_a
    elif operand == 5:
        return reg_b
    elif operand == 6:
        return reg_c
    elif operand == 7:
        print(f'Error! Operand 7')
        return 'Error'

    return


def op0(reg_a, reg_b, reg_c, operand: int):

    numerator = reg_a
    to_power = find_combo(reg_a, reg_b, reg_c, operand)
    denom = 2**to_power

    res1 = int(numerator/denom)

    reg_a = res1

    return reg_a


def op1(reg_a, reg_b, reg_c, operand: int):

    biwise_xor = reg_b ^ operand

    return biwise_xor


def op2(reg_a, reg_b, reg_c, operand: int):

    mod8_combo = find_combo(reg_a, reg_b, reg_c, operand) % 8

    return mod8_combo


def op3(reg_a, reg_b, reg_c, operand: int):

    if reg_a == 0:
        jump = False
        return jump, 0
    else:
        jump = True
        return jump, operand


def op4(reg_a, reg_b, reg_c, operand: int):

    bitwise_xor = reg_b ^ reg_c

    return bitwise_xor


def op5(reg_a, reg_b, reg_c, operand: int):

    mod8_combo = find_combo(reg_a, reg_b, reg_c, operand) % 8

    return mod8_combo


def op6(reg_a, reg_b, reg_c, operand: int):

    numerator = reg_a
    to_power = find_combo(reg_a, reg_b, reg_c, operand)
    denom = 2**to_power

    res1 = int(numerator/denom)

    return res1


def op7(reg_a, reg_b, reg_c, operand: int):

    numerator = reg_a
    to_power = find_combo(reg_a, reg_b, reg_c, operand)
    denom = 2**to_power

    res1 = int(numerator/denom)

    return res1


def next_instruction():

    return 2


def run_process(reg_a, reg_b, reg_c, prog):
    output = []

    instruction_pointer = 0

    while instruction_pointer < len(prog):

        opcode = prog[instruction_pointer]
        next_operand = prog[instruction_pointer+1]

        if opcode == 0:
            reg_a = op0(reg_a, reg_b, reg_c, next_operand)
            # print(f'opcode: {opcode}')
            # print(f'New reg_a {reg_a}')
            instruction_pointer += next_instruction()

        elif opcode == 1:
            reg_b = op1(reg_a, reg_b, reg_c, next_operand)
            instruction_pointer += next_instruction()

        elif opcode == 2:
            reg_b = op2(reg_a, reg_b, reg_c, next_operand)
            # output.append([out2])
            instruction_pointer += next_instruction()

        elif opcode == 3:
            # print(f'Opcode {opcode} and operand {next_operand} ')
            jump, res3 = op3(reg_a, reg_b, reg_c, next_operand)
            if jump:
                instruction_pointer = res3
            else:
                instruction_pointer += next_instruction()

        elif opcode == 4:
            reg_b = op4(reg_a, reg_b, reg_c, next_operand)
            instruction_pointer += next_instruction()

        elif opcode == 5:
            # print(f'Opcode {opcode} and operand {next_operand} ')
            mod8_combo = op5(reg_a, reg_b, reg_c, next_operand)
            output.append([int(i) for i in list(str(mod8_combo))])

            instruction_pointer += next_instruction()

        elif opcode == 6:
            reg_b = op6(reg_a, reg_b, reg_c, next_operand)
            instruction_pointer += next_instruction

        elif opcode == 7:
            reg_c = op7(reg_a, reg_b, reg_c, next_operand)
            instruction_pointer += next_instruction()

    output = [item for sublist in output for item in sublist]
    output = ','.join([str(i) for i in output])

    print(f' Reg A: {reg_a}, Reg B: {reg_b}, Reg C: {reg_c}')
    return output


reg_a, reg_b, reg_c, prog = read_file()


out = run_process(reg_a, reg_b, reg_c, prog)
print(out)
